filter_sarif.py

```diff
@@ -1,147 +1,218 @@
-import sys
+#!/usr/bin/env python3
+"""
+Filter SARIF files based on file, rule and message patterns.
+
+File paths and rule ids are filtered using glob patterns. Message patterns are python regular expressions.
+"""
+
 import argparse
 import json
 import re
+from typing import Any, Dict, Iterable, NoReturn, Optional, Tuple, List
 from globber import match
+import logging
+
+LOG = logging.getLogger(__name__)
 
+sepchar = ':'
+escchar = '\\'
+include_char = '+'
+exclude_char = '-'
+special_chars = (include_char, exclude_char, escchar, sepchar)
 
-def fail(msg):
+
+def fail(msg: str) -> NoReturn:
+    """Prints an error message and exits with a non-zero exit code."""
     print(msg)
-    sys.exit(-1)
+    exit(-1)
 
 
-def match_path_and_rule(path, rule, patterns):
+def match_patterns(path: Optional[str], rule: Optional[str], message: Optional[str],
+                   patterns: Iterable[Tuple[bool, str, Optional[str], Optional[str]]]) -> bool:
+    """Match a path, rule and message against a list of patterns."""
     result = True
-    for s, fp, rp in patterns:
-        if match(rp, rule) and match(fp, path):
-            result = s
+
+    LOG.debug("Matching patterns against %s:%s:%s", path, rule, message)
+
+    for sign, file_pattern, rule_pattern, message_pattern in patterns:
+        LOG.debug("Matching against %s", (sign, file_pattern, rule_pattern, message_pattern))
+        if (path is None or file_pattern is None or
+                match(file_pattern, path)) and (rule is None or rule_pattern is None or match(rule_pattern, rule)) and (
+                    message_pattern is None or message is None or re.search(message_pattern, message)):
```

Comment:
Is the intent to make it possible to provide message_pattern without a rule_pattern? It doesn't seem to be possible at the moment (I haven't tested this out though, just from my understanding of the code)

Reply:
You can just use a wildcard for the rule pattern to do that, can't you?

I'm not totally sure that globs are the best choice for matching on the rule id, since I think `*` won't actually match on `codeql/this-messsage-id` properly, because of the `/` divider!

``` yaml
**/*:*:^This message is all I want$
```

Reply:
I think it requires a `**`. I somehow assumed that you wanted to allow something like `::^This message is all I want$`.

Reply:
I'll test the wildcard necessary and document it

Reply:
I tested this, and created specific `test_globber.py` tests in the `tests` directory, using `pytest`, to document and test the expected behaviour